### PR TITLE
[SPARK-30481][CORE] Integrate event log compactor into Spark History Server

### PR DESCRIPTION
**Pull request by HeartSaVioR**
### What changes were proposed in this pull request?

This patch addresses remaining functionality on event log compaction: integrate compaction into FsHistoryProvider.

This patch is next task of SPARK-30479 (#27164), please refer the description of PR #27085 to see overall rationalization of this patch.

### Why are the changes needed?

One of major goal of SPARK-28594 is to prevent the event logs to become too huge, and SPARK-29779 achieves the goal. We've got another approach in prior, but the old approach required models in both KVStore and live entities to guarantee compatibility, while they're not designed to do so.

### Does this PR introduce any user-facing change?

No.

### How was this patch tested?

Added UT.

### CODE REVIEW

**Comment by HeartSaVioR** [comment]
Only the last commit belongs to this PR. Will remove WIP label once #27164 is merged.

**Comment by HeartSaVioR** [comment]
cc. @vanzin @squito @gaborgsomogyi

File: core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala [hunk 1]
```diff
@@ -569,6 +570,35 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
...
+   * - 2nd (Option[Long]): Some(value) if the method succeeds to try compaction,
+   *   value indicates the last event log index to try compaction. None otherwise.
+   */
+  private def compact(reader: EventLogFileReader): (Boolean, Option[Long]) = {
+    reader.lastIndex match {
+      case Some(lastIndex) =>
+        try {
+          val info = listing.read(classOf[LogInfo], reader.rootPath.toString)
+          if (info.lastIndexToRunCompaction.isEmpty ||
+            info.lastIndexToRunCompaction.get < lastIndex) {
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala]: indent more

File: core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala [hunk 2]
```diff
@@ -569,6 +570,35 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
     }
   }
 
+  /**
+   * Returns a tuple containing two values. Each element means:
+   * - 1st (Boolean): true if the list of event log files are changed, false otherwise.
+   * - 2nd (Option[Long]): Some(value) if the method succeeds to try compaction,
+   *   value indicates the last event log index to try compaction. None otherwise.
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala]: what does it mean to "try compaction"? Does it mean that when this method returns, no compaction was actually done, it was just tried?

The tuple being returned sounds a bit confusing. instead, why not return just an `Option[Long]` telling you both whether compaction ran, and what's the index of the first non-compacted file (or last compacted file, not sure what's being tracked really)?

File: core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala [hunk 3]
```diff
@@ -661,26 +691,33 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
       reader: EventLogFileReader,
       scanTime: Long,
       enableOptimizations: Boolean): Unit = {
+    val rootPath = reader.rootPath
     try {
+      val (shouldReload, lastCompactionIndex) = compact(reader)
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala]: Is there anyway to do compaction as a separate task? Otherwise it seems like this could slow down the creation of the app listing, especially if you're not using disk caching or it's the first time the SHS is creating the listing db.

File: core/src/main/scala/org/apache/spark/internal/config/package.scala [hunk 4]
```diff
@@ -197,8 +197,6 @@ package object config {
 
   private[spark] val EVENT_LOG_ROLLING_MAX_FILES_TO_RETAIN =
     ConfigBuilder("spark.eventLog.rolling.maxFilesToRetain")
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/internal/config/package.scala]: Hmm... this is kind of a history server config, right? Or even more specifically, a `FsHistoryProvider` config.

Might be better to have it grouped with other configs in `History.scala`, and make it a constructor argument in `EventLogFileCompactor`.

(In fact the same applies to `EVENT_LOG_COMPACTION_SCORE_THRESHOLD`. Should that be not `internal()` anymore, also?)

File: core/src/test/scala/org/apache/spark/deploy/history/FsHistoryProviderSuite.scala [hunk 5]
```diff
@@ -1362,6 +1365,110 @@ class FsHistoryProviderSuite extends SparkFunSuite with Matchers with Logging {
...
+
+      // writing event log file 2 - compact the event log file 1 into 1.compact
+      writeEventsToRollingWriter(writer, Seq.empty, rollFile = true)
+      writeEventsToRollingWriter(writer, Seq(SparkListenerUnpersistRDD(1),
+        SparkListenerJobEnd(1, 1, JobSucceeded)), rollFile = false)
+
+      updateAndCheck(provider) { _ =>
+        verifyEventLogFiles(fs, writer.logPath, Some(1), Seq(2))
+        val info = provider.listing.read(classOf[LogInfo], writer.logPath)
+        assert(info.lastIndexToRunCompaction === Some(2))
```
Comment (vanzin) [inline core/src/test/scala/org/apache/spark/deploy/history/FsHistoryProviderSuite.scala]: I'm a little confused by this check. `2` was not compacted, right? Only `1` was. So why is `lastIndexToRunCompaction` 2?

File: docs/configuration.md [hunk 6]
```diff
@@ -1023,6 +1023,24 @@ Apart from these, the following properties are also available, and may be useful
     The max size of event log file before it's rolled over.
   </td>
 </tr>
+<tr>
+  <td><code>spark.eventLog.rolling.maxFilesToRetain</code></td>
+  <td>Int.MaxValue</td>
+  <td>
+    The maximum number of event log files which will be retained as non-compacted.
+    By default, all event log files will be retained. Please set the configuration and
```
Comment (vanzin) [inline docs/configuration.md]: Might be good to clarify where to configure things, since this config is for the history server, but `spark.eventLog.rolling.maxFileSize` needs to be set in the applications.

File: core/src/main/scala/org/apache/spark/internal/config/package.scala (hunk 4 above)
Comment (HeartSaVioR) [reply core/src/main/scala/org/apache/spark/internal/config/package.scala]: > Hmm... this is kind of a history server config, right? Or even more specifically, a FsHistoryProvider config.
> Might be better to have it grouped with other configs in History.scala, and make it a constructor argument in EventLogFileCompactor.

Yeah... the name and place was constructed when I wasn't fully sure where to put, so I just put near neighborhood configuration. But I think it makes sense to move to History, and also add `history` prefix into configuration name to indicate end users that it won't run at driver side. Making it as a part of constructor argument would be also good.

> (In fact the same applies to EVENT_LOG_COMPACTION_SCORE_THRESHOLD. Should that be not internal() anymore, also?)

I'm not 100 % sure who would tune the score manually. Original intention was to make it flexible to ease testing so placed it to configuration and set it to be "internal". If we want to expose this configuration, we should maybe also provide the information on score and the statistics so that end users can see and tune the threshold. WDYT?

File: core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala (hunk 3 above)
Comment (HeartSaVioR) [reply core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala]: I totally agree it's ideal to do so, but it made me thinking the problem too complicated because compaction "modifies" the event log files which "app listing" and "app rebuild" are reading or are having them as a list of files. 

Actually I had to deal with similar thing in `loadDiskStore` and `createInMemoryStore` (there's retry mechanism), but I'm not sure it doesn't get more complicated if we also take app listing into account.

File: core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala (hunk 2 above)
Comment (HeartSaVioR) [reply core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala]: Uh, I tried to differentiate "compacted log index" and "the log index Spark tries to compact", but the words weren't sufficient or appropriate. (I admit it sounds bad naming and maybe also bad explanation but cannot find any better.) It refers latter.

So the reason why we store the log index into LogInfo is to avoid calling `compact` if possible since it's a heavy operation. How?

Given we know how compaction works (especially it excludes the log file of the last index since it may be changing), the result of compaction is idempotent if we provide the same list of event log files.

In other words, once we tried out for certain set of event log files, we don't need to try out again. For example, assuming there're 2.compact, 3, 4 in list of event log files. If we tried out compaction with the list once, regardless of the result (succeed, low score, not enough files), we don't need to try it again, unless we see 5 in the list of event log files.

In fact it's a bit simplified and there're some exceptional cases, like exception happens when compacting, or configurations changed during restart of SHS. Former case is simple, we will fail to store the index into LogInfo anyway, so it should try again in next chance of checking logs. Latter case actually prevents us to leverage the fact, but I'd ignore it as trade-off to gain performance. If we address caching of state in compactor or filter then it may not a big deal to just call compact, but until then I guess we need this.

File: core/src/test/scala/org/apache/spark/deploy/history/FsHistoryProviderSuite.scala (hunk 5 above)
Comment (HeartSaVioR) [reply core/src/test/scala/org/apache/spark/deploy/history/FsHistoryProviderSuite.scala]: Please refer the comment https://github.com/apache/spark/pull/27208#discussion_r367191271 to see what it means. It's intentional to be 2.

File: core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala (hunk 3 above)
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala]: I was thinking that instead of doing this inline here, you could just submit a task to `replayExecutor` after the listing data is updated, so that you give other tasks fetching listing data a chance to run before you try compaction. (You'd call `endProcessing()` at the end of that separate task.)

Any reason why that would not work?

File: core/src/main/scala/org/apache/spark/internal/config/package.scala (hunk 4 above)
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/internal/config/package.scala]: Makes sense to keep the score threshold internal. We can make it public later if there's a good reason for it.

File: core/src/test/scala/org/apache/spark/deploy/history/FsHistoryProviderSuite.scala (hunk 5 above)
Comment (vanzin) [reply core/src/test/scala/org/apache/spark/deploy/history/FsHistoryProviderSuite.scala]: From that comment it seems that what you're saying is that, in this case, the log with index "2" isn't yet compacted, but it's been processed and found that it's not yet worth it to compact it. Is that right?

I think the variable name is confusing because it mixes "last" (which to me evokes something that has happened already) and "to run" (which evokes a future action). So I can't really parse the meaning of the name since it seems to contradict itself.

Comment (vanzin) [reply core/src/test/scala/org/apache/spark/deploy/history/FsHistoryProviderSuite.scala]: If my understanding is correct then perhaps a better name is `lastEvaluatedForCompaction` (didn't add "index" because the name is already pretty long as it is).

Comment (HeartSaVioR) [reply core/src/test/scala/org/apache/spark/deploy/history/FsHistoryProviderSuite.scala]: > From that comment it seems that what you're saying is that, in this case, the log with index "2" isn't yet compacted, but it's been processed and found that it's not yet worth it to compact it. Is that right?

Yes, correct, it applies regardless of the result of compaction.

> If my understanding is correct then perhaps a better name is lastEvaluatedForCompaction (didn't add "index" because the name is already pretty long as it is).

That's a better name. Thanks! Will address.

File: core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala (hunk 3 above)
Comment (HeartSaVioR) [reply core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala]: I guess what you suggest is separating two tasks and applying lock (processing) for each task, especially let all listing tasks submitted first and all compaction tasks submitted later. 

That would work, but we may also want to consider the difference between cleaning logs and compaction - cleaning logs have its own interval and triggered independently, but compaction is triggered conditionally, only when the return value of shouldReload is true. That means we may want to always do compaction for compaction task, irrelevant of status of `processing` - we wouldn't want to skip, but it's not ideal if we make the task be stuck and wait for `processing`. Maybe has to resubmit at the end of task if the task cannot process due to `processing`?

File: core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala [hunk 7]
```diff
@@ -720,6 +691,20 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
...
+        processing(rootPath)
+        val task: Runnable = () => compact(reader)
+        replayExecutor.submit(task)
+      } catch {
+        // let the iteration over the updated entries break, since an exception on
+        // replayExecutor.submit (..) indicates the ExecutorService is unable
+        // to take any more submissions at this time
+        case e: Exception =>
+          logError(s"Exception while submitting task for compaction", e)
+          endProcessing(rootPath)
```
Comment (HeartSaVioR) [inline core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala]: I think checkLog seems to have a bug - if replayExecutor.submit fails and throws an exception, the log is not removed from `processing` which would block processing for the log in both `checkLogs` and `cleanLogs`. 

I'm OK to fix it here as well, or fix it in another PR to avoid mixing up.

File: core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala (hunk 3 above)
Comment (HeartSaVioR) [reply core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala]: Addressed with my own idea in ec3ffd9d37cc68c36918af8682333b22c7b49d3f - please take a look and suggest other idea if you have better idea.

File: core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala (hunk 7 above)
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala]: >  if replayExecutor.submit fails 

That should be extremely rare, but seems simple to handle.

File: core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala [hunk 8]
```diff
@@ -795,15 +800,42 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
...
+
+  private def compact(reader: EventLogFileReader): Unit = {
+    val rootPath = reader.rootPath
+    try {
+      reader.lastIndex match {
+        case Some(lastIndex) =>
+          try {
+            val info = listing.read(classOf[LogInfo], reader.rootPath.toString)
+            if (info.lastEvaluatedForCompaction.isEmpty ||
+              info.lastEvaluatedForCompaction.get < lastIndex) {
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala]: indent more

File: docs/configuration.md [hunk 9]
```diff
@@ -1023,6 +1023,26 @@ Apart from these, the following properties are also available, and may be useful
     The max size of event log file before it's rolled over.
   </td>
 </tr>
+<tr>
+  <td><code>spark.history.fs.eventLog.rolling.maxFilesToRetain</code></td>
+  <td>Int.MaxValue</td>
+  <td>
+    The maximum number of event log files which will be retained as non-compacted. By default,
```
Comment (vanzin) [inline docs/configuration.md]: Spark currently doesn't have a "release notes" document in the docs dir, but I'm wondering if we should mention somewhere that this is a new feature and may not be completely stable (i.e. use with caution, it may delete more data than you expect, cause some UI issues we haven't thought about, etc). There are migration guides, but that doesn't seem like the right place...

File: core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala [hunk 10]
```diff
@@ -795,15 +800,42 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
...
+              // haven't tried compaction for this index, do compaction
+              fileCompactor.compact(reader.listEventLogFiles)
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala]: So one thing that feels a tiny bit odd is that when deciding whether to compact, you're actually considering the last log file, which you won't consider during actual compaction, right?

Wouldn't that cause unnecessary (or too aggressive) compaction at the end of the application, when potentially a bunch of jobs finish and "release" lots of tasks, inflating the compation scoe?

File: core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala (hunk 7 above)
Comment (HeartSaVioR) [reply core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala]: Ah yes that's being handled as I extracted method and addressed there.

File: core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala (hunk 8 above)
Comment (HeartSaVioR) [reply core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala]: I have no idea why it's back to previous. Will fix.

File: docs/configuration.md (hunk 9 above)
Comment (HeartSaVioR) [reply docs/configuration.md]: I think we are using JIRA issue to write content for release note if it's necessary. Labeled as "release-notes" and content in "Docs text".
https://issues.apache.org/jira/browse/SPARK-25016

I'd be pretty much appreciated if you can write the release note content for this issue, as JIRA issue is not good to review and make change continuously.

File: core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala (hunk 10 above)
Comment (HeartSaVioR) [reply core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala]: > So one thing that feels a tiny bit odd is that when deciding whether to compact, you're actually considering the last log file, which you won't consider during actual compaction, right?
> Wouldn't that cause unnecessary (or too aggressive) compaction at the end of the application, when potentially a bunch of jobs finish and "release" lots of tasks, inflating the compation scoe?

That's the intention that callers of compactor don't care about how many files are actually affected. Callers of compactor just need to know that same list of log files would bring same result, unless it fails and throws exception. How many files are excluded in compaction is just a configuration, and the last log file should be excluded is an implementation detail. 

Compactor will ignore the last log file in any way as configured. (1 is the minimum value for max retain log file) So unless the rare case where the log is rolled just before the app is finished, it won't happen.

File: core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala [hunk 11]
```diff
@@ -1175,6 +1237,21 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
     }
     deleted
   }
+
+  /** NOTE: 'task' should ensure it executes 'endProcessing' at the end */
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala]: Hmm, perhaps this method could take care of calling `endProcessing` too (e.g. by wrapping the task)? Should be just a small adjustment at the call site.

File: docs/configuration.md [hunk 12]
```diff
@@ -1023,6 +1023,26 @@ Apart from these, the following properties are also available, and may be useful
     The max size of event log file before it's rolled over.
   </td>
 </tr>
+<tr>
+  <td><code>spark.history.fs.eventLog.rolling.maxFilesToRetain</code></td>
```
Comment (vanzin) [inline docs/configuration.md]: This actually should be with the other SHS configs in `monitoring.md`.

**Review by vanzin — COMMENTED** [review]
Couple of small things. Regarding the release notes, perhaps it would be better to have more explicit documentation about this feature in `monitoring.md`? A separate section in that document would be easier to follow than a lot of text in the config parameters table.

(Feel free to do that as a separate task.)

File: core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala (hunk 11 above)
Comment (HeartSaVioR) [reply core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala]: `finally` statement in `mergeApplicationListing` makes it be complicated, because we should handle the reentrance of lock.

If we move endProcessing from `finally` statement in `mergeApplicationListing` to the end of task here, `processing` on the compaction task would be executed earlier than calling `endProcessing` on the listing task. Marking lock from compaction task would succeed but effectively no-op, and releasing lock from listing task would remove the mark for compaction task as well, which makes the compaction task run without proper lock.

So either we need to make lock much smarter, or document the requirement on caller side. I'm feeling that former one is more complicated than latter one.

**Comment by HeartSaVioR** [comment]
> Regarding the release notes, perhaps it would be better to have more explicit documentation about this feature in monitoring.md? A separate section in that document would be easier to follow than a lot of text in the config parameters table.
> (Feel free to do that as a separate task.)

Yeah I agree that adding all the explanation about the feature into the description of one configuration isn't ideal. As we may have couple of months (at least a month) to address docs/minors after feature freeze, I'd like to deal with it via separate PR. Thanks for the suggestion!

File: core/src/test/scala/org/apache/spark/deploy/history/FsHistoryProviderSuite.scala [hunk 13]
```diff
@@ -1362,6 +1365,110 @@ class FsHistoryProviderSuite extends SparkFunSuite with Matchers with Logging {
...
+        intercept[NoSuchElementException] {
+          appStore.job(1)
+        }
+
+        // but other events should be available even they were in original files to compact
+        val appInfo = appStore.applicationInfo()
+        assert(appInfo.id === "app")
+        assert(appInfo.name === "app")
+
+        // all events in retained file should be available, even they're related to finished jobs
```
Comment (vanzin) [inline core/src/test/scala/org/apache/spark/deploy/history/FsHistoryProviderSuite.scala]: all execs? (also, should be "all live execs"?)

**Review by vanzin — COMMENTED** [review]
Small comment nit, otherwise looks ok.

`FsHistoryProvider.scala` is getting pretty large and hard to follow, at some point it might be good to break it down and move things like UI replay logic, cleaning and compaction out of it...

File: core/src/test/scala/org/apache/spark/deploy/history/FsHistoryProviderSuite.scala (hunk 13 above)
Comment (HeartSaVioR) [reply core/src/test/scala/org/apache/spark/deploy/history/FsHistoryProviderSuite.scala]: Uh, even the latter verification code technically checks the events for executors (as it's relatively easier to check with AppStatusStore), the comment is still correct - we don't filter out any events in "retained" file(s). I'm refining the comment a bit to clarify these events will include removed executors as well.

**Comment by HeartSaVioR** [comment]
> FsHistoryProvider.scala is getting pretty large and hard to follow, at some point it might be good to break it down and move things like UI replay logic, cleaning and compaction out of it...

Totally agreed. It would be also good for FsHistoryProviderSuite to follow the change of FsHistoryProvider which may be the chance to reduce down its size as well.

**Comment by vanzin** [comment]
Alright, merging to master.

**Comment by HeartSaVioR** [comment]
Finally! Thanks again @vanzin for shepherding the feature and reviewing thoughtfully.

BTW I might start to handle the follow-up TODO in couple of days later (maybe next week?) - stuck with some work.
